### PR TITLE
Detect service from url

### PR DESCRIPTION
**Pull request by pape-io**
When using `includes` to include linked objects use known services for doing the request to be able to apply endpoint configurations from their service definitions (e.g. Auth).

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -9,3 +9,5 @@ gemspec
 # your gemspec. These might include edge Rails or gems from your path or
 # Git. Remember to move these dependencies to your gemspec before releasing
 # your gem to rubygems.org.
+
+gem 'lhc', '~> 1.2.0', path: '../lhc'
```
Comment (skaestle) [inline Gemfile]: local path is referenced

File: lib/lhs.rb [hunk 2]
```diff
@@ -4,3 +4,10 @@ module LHS
 end
 
 Gem.find_files('lhs/**/*.rb').each { |path| require path }
+
+# Preload all the services that are defined in app/services
+class Engine < Rails::Engine
+  initializer 'Load all services' do |app|
+    Dir.glob(app.root.join('app/services/**.rb')).each {|file| require file }
```
Comment (skaestle) [inline lib/lhs.rb]: maybe you wanna include sub directories?

File: lib/lhs/concerns/service/endpoints.rb [hunk 3]
```diff
@@ -18,6 +19,15 @@ def endpoint(url, options = nil)
         endpoint = LHC::Endpoint.new(url, options)
         instance.sanity_check(endpoint)
         instance.endpoints.push(endpoint)
+        LHS::Service::Endpoints.all ||= {}
```
Comment (skaestle) [inline lib/lhs/concerns/service/endpoints.rb]: can you initialize it in the initializer method?

Comment (pape-io) [inline lib/lhs/concerns/service/endpoints.rb]: initalizer method for a module?

File: lib/lhs/concerns/service/request.rb [hunk 4]
```diff
@@ -15,38 +15,34 @@ def request(options)
...
-    end
-
-    def single_request(options)
-      response = LHC.request(process_options(options))
-      data = LHS::Data.new(response.body, nil, self.class, response.request)
-      handle_includes(data) if includes
-      data
+    # Convert URLs in options to endpoint templates
+    def convert_options_to_endpoints(options)
+      if options.is_a? Array
```
Comment (skaestle) [inline lib/lhs/concerns/service/request.rb]: i would do: `options.respond_to?(:map)`

File: lib/lhs/concerns/service/endpoints.rb (hunk 3 above)
Comment (skaestle) [inline lib/lhs/concerns/service/endpoints.rb]: agreed to leave it as it is

File: lib/lhs/concerns/service/request.rb [hunk 5]
```diff
@@ -59,36 +55,77 @@ def handle_includes(data)
...
+      params.delete(:params)
+      params.merge!(explicit_params) if explicit_params
+    end
+
+    def multiple_requests(options)
+      options = options.map { |options| process_options(options) }
+      responses = LHC.request(options)
+      data = responses.map{ |response| LHS::Data.new(response.body, nil, self.class, response.request) }
+      data = LHS::Data.new(data, nil, self.class)
+      handle_includes(data) if includes
```
Comment (skaestle) [inline lib/lhs/concerns/service/request.rb]: `self.` should be superfluous

File: lib/lhs/concerns/service/request.rb [hunk 6]
```diff
@@ -59,36 +55,77 @@ def handle_includes(data)
...
+      options = (endpoint.options || {}).merge(options)
+      options[:url] = compute_url!(options[:params]) unless options.key?(:url)
+      merge_explicit_params!(options[:params])
+      options.delete(:params) if options[:params] && options[:params].empty?
+      options
+    end
+
+    def service_for_options(options)
+      services = []
+      if options.is_a?(Array)
```
Comment (skaestle) [inline lib/lhs/concerns/service/request.rb]: again, imo: `options.respond_to?(:each)`

**Comment by skaestle** [comment]
documentation needs update, but other than that: :+1:

File: lib/lhs/concerns/service/request.rb (hunk 5 above)
Comment (pape-io) [inline lib/lhs/concerns/service/request.rb]: http://ruby-doc.org/core-2.2.0/Class.html 
69ee5fbbec1cc6d10a91fbdb34ac5503

**Pull request merged**